Update doc builder recommendations. (#327)
UCL-ARC/python-tooling: docs/pages/docs.md

# Documentation

With Python, as with many other languages, it's very common to automatically
create a web page with documentation built from [docstrings] in your package's

> API. For your package to be easily reusable we recommend you build documentation

Reviewer: ```suggestion
_application programming interface_ (API). For your package to be easily reusable we recommend you build documentation
```
Possibly overkill but even for commonly used acronyms I think its worth defining in full on first usage though we didn't define API in its usage before these edits so feel free to ignore

> With Python, as with many other languages, it's very common to automatically
> create a web page with documentation built from [docstrings] in your package's
> API. For your package to be easily reusable we recommend you build documentation

Reviewer 2: ```suggestion
With Python, as with many other languages, it's very common to automatically
create a web page with documentation. This can include reference for the API taken
from your package's [docstrings], tutorials, examples, and more in depth explanation
about how the package works. For an in-depth discussion of the different components
of good documentation, see the [Diátaxis framework](https://diataxis.fr/).

For your package to be easily reusable we recommend you build documentation
```
Since API docs are not the only type of docs?

Author: Could not apply suggestion with the nice button. [c97ca4e](https://github.com/UCL-ARC/python-tooling/pull/327/commits/c97ca4edcde8990a6fb35cb104201046d87ad480)

pages and host them somewhere.

> If you're using GitHub, the simplest is to host your docs on [GitHub pages].

Reviewer 2: I would disagree, and in my opinion it's simpler to use [readthedocs](https://readthedocs.org/) - would suggest we remove this in this PR, and defer documentation hosting dicsussion to a new issue?

Author: Oh! I never used RTD, but just saw you fighting with it for napari-matplotlib. It's what we do in the template. How about `s/simplest is to/one option is to/` ?

Author: ```suggestion
If you're using GitHub, one option is to host your docs on [GitHub pages].
```

Reviewer 2: 👍  - `napari-matplotlib` was very much the outlier of "doesn't work with readthedocs" because we were trying to include 3D rendered images in the docs, and generate them as part of the build.

[Here's how we've set this up][template-docs-dot-yaml] for [our template].

<!-- URL used above in the blurb-->

[docstrings]: https://peps.python.org/pep-0257/#what-is-a-docstring
[GitHub pages]: https://docs.github.com/en/pages
[our template]: https://github.com/UCL-ARC/python-tooling?tab=readme-ov-file#using-this-template
[template-docs-dot-yaml]: https://github.com/UCL-ARC/python-tooling/blob/main/%7B%7Bcookiecutter.project_slug%7D%7D/.github/workflows/docs.yml

## Documentation build tools

| Name      | Short description                                                                                                                                               | 🚦  |
| --------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------- | :-: |

> | [MkDocs]  | Generates documentation from Markdown files, with a plugin to generate API documentation. A good plugin ecosystem and balance of usability and customisability. | 🟢  |

Reviewer 2: Add this plugin to a new section for mkdocs plugins below?

Author: - [x] Add section for MkDocs plugins.

| [Sphinx]  | Generates documentation from reStructuredText or Markdown files, long been the de-facto standard and very widely used. Mature plugin ecosystem.                 | 🟠  |
| [gitbook] | General documentation builder; integrates with GitHub.                                                                                                          | 🟠  |
| [pdoc]    | Auto-generates API documentation from docstrings, beginner friendly but with less of a plugin ecosystem than others.                                            | 🟠  |

<details markdown="block">
<summary>More details about Sphinx</summary>

We marginally recommend [MkDocs] over [Sphinx] due to it's ease of use,

> preference for markdown, and slightly better support of our preferred docstring

Reviewer: ```suggestion
preference for Markdown, and slightly better support of our preferred docstring
```
Nitpick for consistency with capitalisation elsewhere

> preference for markdown, and slightly better support of our preferred docstring

Reviewer 2: What is our preferred docstring style? Again, unless I missed it (sorry), I think worth deferring discussion/choice of a docstring style to a new issue?

Author: P'raps: "more support for a variety of docstring styles" ...?

Reviewer: Agree with having a separate discussion about docstring style and rephrasing as @samcunliffe suggested

Author: - [x] ☝️

style.

> However the [Sphinx] tool has long been the de-facto standard, it is widely

Reviewer 2: ```suggestion
However the [Sphinx] tool has long been the de-facto standard in the scientific Python ecosystem. It is widely
```

used, customisable, and well tested. If you need a [Sphinx
extension](#sphinx-extensions) that does not have an equivalent [MkDocs
plugin](https://github.com/mkdocs/catalog), or if you are part of a community

> that heavily uses [Sphinx] then, as with all tools, we recommend you use that

Reviewer 2: ```suggestion
that heavily uses [Sphinx] then we recommend you use that
```

instead.

### See also

> - Our internal discussions about which to recommend ([#16](https://github.com/UCL-ARC/python-tooling/issues/16) and [#187](https://github.com/UCL-ARC/python-tooling/issues/187)).

Reviewer: Nice 😁 - I think linking to discussions like this is a good shout as it allows users to see the rationales behind our recommendations (and that we have actually thought about them!).

<details markdown="block">
<summary>More details about Sphinx</summary>

We marginally recommend [MkDocs] over [Sphinx] due to it's ease of use,
preference for Markdown, and more support for a variety of docstring styles.

However the [Sphinx] tool has long been the de-facto standard in the scientific Python ecosystem. It is widely
used, customisable, and well tested. If you need a [Sphinx
extension](#sphinx-extensions) that does not have an equivalent [MkDocs
plugin](https://github.com/mkdocs/catalog), or if you are part of a community
that heavily uses [Sphinx] then we recommend you use that
instead.

### See also

- Our internal discussions about which to recommend ([#16](https://github.com/UCL-ARC/python-tooling/issues/16) and [#187](https://github.com/UCL-ARC/python-tooling/issues/187)).
- [An interesting related discussion](https://github.com/encode/httpx/discussions/1220).

</details>

<!-- URLS used above -->

[MkDocs]: https://www.mkdocs.org/
[Sphinx]: https://www.sphinx-doc.org/en/master/
[gitbook]: https://www.gitbook.com/
[pdoc]: https://pdoc.dev/

## MkDocs plugins

> | name                                                          | short description                            | 🚦  |

Author: ```suggestion
| Name                                                          | Short description                            | 🚦  |
```

| ------------------------------------------------------------- | -------------------------------------------- | :-: |

> | [mkdocstrings-python](https://mkdocstrings.github.io/python/) | Automatically generates api reference pages. | 🟢  |

Reviewer: ```suggestion
| [mkdocstrings-python](https://mkdocstrings.github.io/python/) | Automatically generates API reference pages. | 🟢  |
```

Author: Thanks. This is me typing too fast, I think.

## Sphinx extensions

> | name                                                                 | short description                                                | 🚦  |

Author: ```suggestion
| Name                                                                 | Short description                                                | 🚦  |
```

| -------------------------------------------------------------------- | ---------------------------------------------------------------- | :-: |

> | [sphinx-gallery](https://sphinx-gallery.github.io/stable/index.html) | Builds an html gallery of examples from a set of python scripts. | 🟢  |

Reviewer: ```suggestion
| [sphinx-gallery](https://sphinx-gallery.github.io/stable/index.html) | Builds an HTML gallery of examples from a set of Python scripts. | 🟢  |
```

> | [sphinx-autoapi](https://sphinx-autoapi.readthedocs.io/en/stable/)   | Automatically generates api reference pages.                     | 🟢  |

Reviewer: ```suggestion
| [sphinx-autoapi](https://sphinx-autoapi.readthedocs.io/en/stable/)   | Automatically generates API reference pages.                     | 🟢  |
```

Author: Huh, something's gone wrong with capitalisation. Sorry for the fail/noise 🙃 
```
| name                                                                 | short description                                                | 🚦  |
```
